### PR TITLE
rpc: fix data optionality for RPC calls.

### LINKED ISSUE #27828: bitcoin-cli output help text - (json object) not utilizing RPCArg::Optional value
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behaviour

When executing: `bitcoin-cli walletcreatefundedpsbt` or `bitcoin-cli createpsbt` without args you will get a help output that looks like this:
```
Creates and funds a transaction in the Partially Signed Transaction format.
Implements the Creator and Updater roles.
All existing inputs must either have their previous output transaction be in the wallet
or be in the UTXO set. Solving data must be provided for non-wallet inputs.

Arguments:
1. inputs                             (json array, optional) Leave empty to add inputs automatically. See add_inputs option.
     [
       {                              (json object)
         "txid": "hex",               (string, required) The transaction id
         "vout": n,                   (numeric, required) The output number
         "sequence": n,               (numeric, optional, default=depends on the value of the 'locktime' and 'options.replaceable' arguments) The sequence number
         "weight": n,                 (numeric, optional, default=Calculated from wallet and solving data) The maximum weight for this input, including the weight of the outpoint and sequence number. Note that signature sizes are not guaranteed to be consistent, so the maximum DER signatures size of 73 bytes should be used when considering ECDSA signatures.Remember to convert serialized sizes to weight units when necessary.
       },
       ...
     ]
2. outputs                            (json array, required) The outputs (key-value pairs), where none of the keys are duplicated.
                                      That is, each address can only appear once and there can only be one 'data' object.
                                      For compatibility reasons, a dictionary, which holds the key-value pairs directly, is also
                                      accepted as second parameter.
     [
       {                              (json object)
         "address": amount,           (numeric or string, required) A key-value pair. The key (string) is the bitcoin address,
                                      the value (float or string) is the amount in BTC
         ...
       },
       {                              (json object)
         "data": "hex",               (string, required) A key-value pair. The key must be "data", the value is hex-encoded data
       },
       ...
     ]
3. locktime                           (numeric, optional, default=0) Raw locktime. Non-0 value also locktime-activates inputs
4. options                            (json object, optional)
     {
       "add_inputs": bool,            (boolean, optional, default=false when "inputs" are specified, true otherwise) Automatically include coins from the wallet to cover the target amount.
                                      
       "include_unsafe": bool,        (boolean, optional, default=false) Include inputs that are not safe to spend (unconfirmed transactions from outside keys and unconfirmed replacement transactions).
                                      Warning: the resulting transaction may become invalid if one of the unsafe inputs disappears.
                                      If that happens, you will need to fund the transaction with different inputs and republish it.
       "minconf": n,                  (numeric, optional, default=0) If add_inputs is specified, require inputs with at least this many confirmations.
       "maxconf": n,                  (numeric, optional) If add_inputs is specified, require inputs with at most this many confirmations.
       "changeAddress": "str",        (string, optional, default=automatic) The bitcoin address to receive the change
       "changePosition": n,           (numeric, optional, default=random) The index of the change output
       "change_type": "str",          (string, optional, default=set by -changetype) The output type to use. Only valid if changeAddress is not specified. Options are "legacy", "p2sh-segwit", "bech32", and "bech32m".
       "includeWatching": bool,       (boolean, optional, default=true for watch-only wallets, otherwise false) Also select inputs which are watch only
       "lockUnspents": bool,          (boolean, optional, default=false) Lock selected unspent outputs
       "fee_rate": amount,            (numeric or string, optional, default=not set, fall back to wallet fee estimation) Specify a fee rate in sat/vB.
       "feeRate": amount,             (numeric or string, optional, default=not set, fall back to wallet fee estimation) Specify a fee rate in BTC/kvB.
       "subtractFeeFromOutputs": [    (json array, optional, default=[]) The outputs to subtract the fee from.
                                      The fee will be equally deducted from the amount of each specified output.
                                      Those recipients will receive less bitcoins than you enter in their corresponding amount field.
                                      If no outputs are specified here, the sender pays the fee.
         vout_index,                  (numeric) The zero-based output index, before a change output is added.
         ...
       ],
       "conf_target": n,              (numeric, optional, default=wallet -txconfirmtarget) Confirmation target in blocks
       "estimate_mode": "str",        (string, optional, default="unset") The fee estimate mode, must be one of (case insensitive):
                                      "unset"
                                      "economical"
                                      "conservative"
       "replaceable": bool,           (boolean, optional, default=wallet default) Marks this transaction as BIP125-replaceable.
                                      Allows this transaction to be replaced by a transaction with higher fees
       "solving_data": {              (json object, optional) Keys and scripts needed for producing a final transaction with a dummy signature.
                                      Used for fee estimation during coin selection.
         "pubkeys": [                 (json array, optional, default=[]) Public keys involved in this transaction.
           "pubkey",                  (string) A public key
           ...
         ],
         "scripts": [                 (json array, optional, default=[]) Scripts involved in this transaction.
           "script",                  (string) A script
           ...
         ],
         "descriptors": [             (json array, optional, default=[]) Descriptors that provide solving data for this transaction.
           "descriptor",              (string) A descriptor
           ...
         ],
       },
     }
5. bip32derivs                        (boolean, optional, default=true) Include BIP 32 derivation paths for public keys if we know them

Result:
{                     (json object)
  "psbt" : "str",     (string) The resulting raw transaction (base64-encoded string)
  "fee" : n,          (numeric) Fee in BTC the resulting transaction pays
  "changepos" : n     (numeric) The position of the added change output, or -1
}

Examples:

Create a transaction with no inputs
> bitcoin-cli walletcreatefundedpsbt "[{\"txid\":\"myid\",\"vout\":0}]" "[{\"data\":\"00010203\"}]"
```

The (json object)'s within `outputs` are not being populated with "required" or "optional"

### Expected behaviour

`(json object, optional)` or `(json object, required)` appears in help text. 
Edit: This is more complicated as 1 of the 2 are minimally needed. Instead helper text was added.

### Steps to reproduce

Running v25 of bitcoin core and using bitcoin-cli.

### Relevant log output

_No response_

### How did you obtain Bitcoin Core

Pre-built binaries

### What version of Bitcoin Core are you using?

v25.0

### Operating system and version

Ubuntu 22 LTS

### Machine specifications

_No response_

### PR DESCRIPTION
**Pull request by miketwenty1**
The "data" field without outputs was marked as "required" in the help docs when using bitcoin-cli. This field when left off worked as an intended optional OP_RETURN. closes #27828.

Motivation: It is hard to understand that "data" is actually optional for commands like `createpsbt` and `walletcreatefundedpsbt`.

### CODE REVIEW

**Comment by miketwenty1** [comment]
> I get that this is confusing, but this change isn't correct.
> 
> The `"data"` field itself is required when the object is present (as in `[{"address": amount, ..., {}]` would not be valid). What's optional is the object (the `{ ... }` around it), which is at a higher level.
> 
> Maybe it's worth elaborating in the prose text above?

Understood you're saying the change isn't correct. How would someone properly indicate the "data" is optional if not specifying it as optional? The parent output is required and some of it's children elements are optional. From my perspective what I would expect to see is a required "outputs" and within outputs there being required and optional fields. 

For instance: `"address": amount,           (numeric or string, required` makes sense as address and amount are necessary.
But `"data": "hex",               (string, required)` seems like a bug because this is not required and can be left out. If you specify "data" then of course there will be a required value for it, but this is implied with how json works.

Full text of resulting text for  outputs as reference:
```
2. outputs                            (json array, required) The outputs (key-value pairs), where none of the keys are duplicated.
                                      That is, each address can only appear once and there can only be one 'data' object.
                                      For compatibility reasons, a dictionary, which holds the key-value pairs directly, is also
                                      accepted as second parameter.
     [
       {                              (json object)
         "address": amount,           (numeric or string, required) A key-value pair. The key (string) is the bitcoin address,
                                      the value (float or string) is the amount in BTC
         ...
       },
       {                              (json object)
         "data": "hex",               (string, required) A key-value pair. The key must be "data", the value is hex-encoded data
       },
       ...
     ]
```

**Comment by miketwenty1** [comment]
> The `"data": ...` part cannot be left out. It's the `{"data": ...}` that can be left out.

Let me rephrase my question:
How would someone know `{"data" ...}` is optional if not specifying it as optional? If every field is optional within an obj maybe it's implied the obj itself is optional?

Does there need to be a new idea of specifying whether the (json object) itself is optional?

**Comment by miketwenty1** [comment]
@sipa, edited the commit. Something like this look good?

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -149,6 +149,7 @@ static std::vector<RPCArg> CreateTxDoc()
         },
         {"outputs", RPCArg::Type::ARR, RPCArg::Optional::NO, "The outputs (key-value pairs), where none of the keys are duplicated.\n"
                 "That is, each address can only appear once and there can only be one 'data' object.\n"
+                "Either {\"data\"} or {\"address\"}, are required, but specifying both is optional.\n"
```
Comment (Sjors) [inline src/rpc/rawtransaction.cpp:152]: I find this wording confusing. Do you mean to say that the user can combine data and address entries? Or that you can provide data or an address directly without putting "data" or "address" in front of it?

Examples can be helpful. Maybe say "see send RPC for examples" and then expand the `send` RPC docs with said examples.

Comment (miketwenty1) [reply src/rpc/rawtransaction.cpp:152]: > Do you mean to say that the user can combine data and address entries? 

The outputs array, `outputs []`, must contain at minimum either an obj containing `{"address" ".."}` or an obj containing `{"data": ".."}`. You can also specify both, but you must have at least one.

I'm reading over this block of help text again. 
```
(json array, required) The outputs (key-value pairs), where none of the keys are duplicated.
                                      That is, each address can only appear once and there can only be one 'data' object.
                                      For compatibility reasons, a dictionary, which holds the key-value pairs directly, is also
                                      accepted as second parameter.
```

just using separate objs
```json 
[
    {
        "bc1address1":"0.01",
    },
    {
        "bc1address2":"0.02",
    },
    {
        "data":"ff00ff00ff"
    }
]
```
or...
Also note:
`a dictionary, which holds the key-value pairs directly, is also accepted as second parameter.`
I believe this is saying you can also do something like:
```json 
[
    {
        "bc1address1":"0.01",
        "bc1address2":"0.02",
        "data":"ff00ff00ff"
    }
]
```
I think this style is for compatibility reasons per docs, so I'm guessing it's preferred/future proof to break everything out into individual objects. Please correct me if I'm wrong!

**Comment by miketwenty1** [comment]
@Sjors Not sure if you saw my response above, let me know what you think, we can close this out if there's no appetite for this small change.

File: src/rpc/rawtransaction.cpp (hunk 1 above)
Comment (murchandamus) [inline src/rpc/rawtransaction.cpp:152]: How about something along these lines?

```suggestion
        {"outputs", RPCArg::Type::ARR, RPCArg::Optional::NO, "The transaction’s outputs specified as key-value pairs.\n"
                 "Payment outputs are specified in the form of \"address\":\"amount [BTC]\", e.g. \"bc1p…xyz\":\"0.001\".\n"
                 "A single data output may be specified in the form of \"data\":\"hex\", e.g. \"data\":\"ff00ff00ff\".\n"
                 "Each key may only appear once. This means that each address may only appear once, and there can only be one 'data' output.\n"
                 "At least one output must be specified.\n"
```

**Comment by murchandamus** [comment]
> @Sjors Not sure if you saw my response above, let me know what you think, we can close this out if there's no appetite for this small change.

I think you have correctly identified that this text could be improved and we should move forward with the PR.

**Comment by miketwenty1** [comment]
@jonatack thanks for the bump. Open for feedback. I took @murchandamus suggestions where it seemed appropriate.

**Comment by Sjors** [comment]
I think the [CI failure](https://cirrus-ci.com/task/6512353575960576) is unrelated to your change (see #28027).

**Review by Sjors — APPROVED** [review]
tACK 27b168b81f3959f5376a4176673187a71c7b25e1

I didn't think about it very deeply, but the new text is an improvement in clarity.

I think the [CI failure](https://cirrus-ci.com/task/6512353575960576) is unrelated to your change (see #28027, cc @achow101).

**Comment by achow101** [comment]
ACK 27b168b81f3959f5376a4176673187a71c7b25e1
